### PR TITLE
Implement finalizers

### PR DESCRIPTION
**Pull request by mystor**
This allows us to reduce our dependence on Drop for destroying values
stored in the Gc heap.

I wrote this prototype implementation a while ago - I _think_ it was finished (tests are passing now that I come back to it, and the code looks sane-ish).

r? @Manishearth

### CODE REVIEW

File: gc/src/trace.rs [hunk 1]
```diff
@@ -1,5 +1,18 @@
+/// The Finalize trait. Can be specialized for a specific type to define
+/// finalization logic for that type.
+pub trait Finalize {
+    fn finalize(&self);
```
Comment (apasel422) [inline gc/src/trace.rs]: Is there a reason to have this method in its own trait instead of making it a defaulted method of `Trace`?

Comment (mystor) [inline gc/src/trace.rs]: Basically because the user isn't defining the `Trace` definition for their type manually, they are supposed to `#[derive(Trace)]`, which means that they don't have the ability to specify finalize manually. Even with specialization, you can't have two `impl Trace for T {...}` blocks, one of which would be generated by the `derive` plugin, and the other would be written by the user which wants to implement `Finalize`.

Basically, if we were to put it on Trace, and someone was to use the plugin to implement trace, then they couldn't also manually implement Finalize, as their implementation would conflict.

For example: https://is.gd/w112yQ

**Comment by arthurprs** [comment]
[possibly stupid question]

What's the advantage of having this in addition to Drop?

**Comment by Manishearth** [comment]
Drop itself is unsafe to implement on GC types.

**Comment by arthurprs** [comment]
I see, because the order of drops are non-deterministic?

**Comment by Manishearth** [comment]
Sort of, also that while sweeping drop may try to access a GCd field that has already been swept
